apa102: fix end frame

fouge opened the pull request:
end frame is used to supply clock pulses so that data goes to next led. 
It depends on the number of LEDs in the chain.

Previously, the number of ones sent into the end frame was hard-coded and limited the usage of the driver to 64 LEDs in the strip.
This PR fixes that issues by computing the number of bits to send in the end frame

simonguinot commented on drivers/led_strip/apa102.c:
```diff
@@ -21,7 +21,17 @@ static int apa102_update(const struct device *dev, void *buf, size_t size)
+	 * number of clock pulses required is exactly half the total number
+	 * of LEDs in the string
+	 */
+	uint8_t ones[((size / sizeof(struct led_rgb)) / sizeof(uint8_t) / 2) + 1];
```
`/ sizeof(uint8_t)` is a division by one. What purpose does it serve ?

simonguinot commented on drivers/led_strip/apa102.c:
```diff
@@ -50,7 +60,8 @@ static int apa102_update(const struct device *dev, void *buf, size_t size)
 
-static int apa102_update_rgb(const struct device *dev, struct led_rgb *pixels,
+static int apa102_update_rgb(const struct device *dev,
+			     struct led_rgb *pixels,
```
Please, revert this cosmetic change. It is not related with the current patch. As a rule of thumb, don't pollute a patch with unrelated changes.

simonguinot commented on drivers/led_strip/apa102.c:
```diff
@@ -98,6 +109,7 @@ static const struct led_strip_driver_api apa102_api = {
 };
 
 #define APA102_DEVICE(idx)						 \
+	BUILD_ASSERT(sizeof(struct led_rgb) == 4);			 \
```
Is that to ensure that `LED_STRIP_RGB_SCRATCH` is enabled ? Or something else ?

simonguinot reviewed (changes requested):
Hi @fouge,

Thanks for this patch. Please see my comments below.

Additionally, can you please improve the commit description message ? The PR description message you wrote is more detailed and understandable. You could use it.

simonguinot commented:
ping @fouge :)

fouge commented:
@simonguinot sorry for the very late reply, & thank you for the review.
i addressed all your comments.

fouge committed:
apa102: fix end frame

end frame is used to supply clock pulses so that data goes to last
LED in the chain. Thus, it depends on the number of LEDs in the chain.
Previously, the number of ones sent into the end frame was
hard-coded and limited the usage of the driver to 64 LEDs in the
strip.

Signed-off-by: Cyril Fougeray <cyril.fougeray@toolsforhumanity.com>

simonguinot commented on drivers/led_strip/apa102.c:
```diff
@@ -21,7 +21,17 @@ static int apa102_update(const struct device *dev, void *buf, size_t size)
 {
 	const struct apa102_config *config = dev->config;
 	static const uint8_t zeros[] = { 0, 0, 0, 0 };
```
And is that OK to have only 32 zeros if there are more than 64 LEDs ?

fouge replied:
yes

simonguinot replied:
OK ! :)

simonguinot commented:
> @simonguinot sorry for the very late reply, & thank you for the review. i addressed all your comments.

No problem. It looks good. Please see above for a last question.

simonguinot reviewed (dismissed):
Thanks @fouge for this bug fix !

simonguinot commented:
@thedjnK @soburi Please have a look

fabiobaltieri commented on drivers/led_strip/apa102.c:
```diff
@@ -21,7 +21,17 @@ static int apa102_update(const struct device *dev, void *buf, size_t size)
+	 * number of clock pulses required is exactly half the total number
+	 * of LEDs in the string
+	 */
+	uint8_t ones[((size / sizeof(struct led_rgb)) / 2) + 1];
```
Wondering if it is really a good idea to allocate a dynamically sized array on the stack, especially since it's just filled with 0xff, this could waste an awful lot of RAM on a long strip.

How about allocating an 0xff array statically up to a certain configurable size? The drawback is then that the maximum strip length would have to be configured at build time but imo that's way better than changing this as potentially blowing up the stack

simonguinot replied:
Hi @fabiobaltieri. Thanks for spotting this.

@fouge Please can you statically `initialize` the ones array (with the half the strip length) and pass it via the `dev->data` pointer ?

fabiobaltieri replied:
Can't say I haven't done it myself. :-)

fouge replied:
I could make the array `const` with some macro that are available with GNU extension
https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html#Designated-Initializers

```
uint8_t apa102_end_frame_##idx[] = { [0 ... LENGTH] = 0xFF };
```

not sure we want a specific implementation for GNU though, what do you think?

fabiobaltieri replied:
wouldn't risk it, the project as a whole aims to be compatible with other compilers that we don't have access to in CI, not a fan of that myself but that's unfortunately something we have to deal with

fabiobaltieri reviewed (commented):
One comment on the allocation. Also the commit message can be 72 columns long, no need to make it look like a haiku :-)

simonguinot reviewed (approved):
Well done !
